packages/toolpad-app/src/toolpad/AppEditor/StructureExplorer/index.tsx

```diff
@@ -0,0 +1,286 @@
+    );
+  }
+  if (renderItem.length) {
+    return (
+      <TreeItem nodeId={root.id} label={<Typography variant="body2">{root.name}</Typography>}>
+        <TreeItem
+          nodeId={`${root.id}-renderItem`}
+          label={<Typography variant="body2">renderItem</Typography>}
+        >
+          {renderItem.map((childNode) => (
+            <RecursiveSubTree
+              key={childNode.id}
+              dom={dom}
+              root={childNode}
+              onHover={onHover}
+              onMouseLeave={onMouseLeave}
+            />
+          ))}
+        </TreeItem>
+      </TreeItem>
+    );
+  }
+
+  return (
+    <CustomTreeItem nodeId={root.id} node={root} onHover={onHover} onMouseLeave={onMouseLeave} />
+  );
+}
+
+export default function PageStructureExplorer() {
+  const { dom } = useDom();
+  const { currentView } = useAppState();
+  const appStateApi = useAppStateApi();
+  const [expandedDomNodeIds, setExpandedDomNodeIds] = React.useState<string[]>([]);
+
+  const currentPageId = currentView?.kind === 'page' ? currentView.nodeId : null;
+  const currentPageNode = currentPageId ? appDom.getNode(dom, currentPageId, 'page') : null;
+  const selectedDomNodeId = currentView.kind === 'page' ? currentView.selectedNodeId : '';
+
+  const selectedNodeAncestorIds = React.useMemo(() => {
+    if (!selectedDomNodeId) {
+      return [];
+    }
+    const selectedNode = appDom.getMaybeNode(dom, selectedDomNodeId);
+    if (selectedNode) {
+      return appDom.getAncestors(dom, selectedNode).map((node) => node.id);
+    }
+    return [];
+  }, [dom, selectedDomNodeId]);
+
+  const { children: rootChildren = [] } = React.useMemo(() => {
+    if (!currentPageNode) {
+      return { children: [] };
+    }
+    return appDom.getChildNodes(dom, currentPageNode);
+  }, [dom, currentPageNode]);
+
+  const handleNodeSelect = React.useCallback(
+    (event: React.SyntheticEvent, nodeId: string) => {
+      appStateApi.selectNode(nodeId as NodeId);
+    },
+    [appStateApi],
+  );
+
+  const handleNodeFocus = React.useCallback(
+    (event: React.SyntheticEvent, nodeId: string) => {
+      appStateApi.hoverNode(nodeId as NodeId);
+    },
+    [appStateApi],
+  );
+
+  const handleNodeHover = React.useCallback(
+    (event: React.MouseEvent, nodeId: NodeId) => {
+      appStateApi.hoverNode(nodeId as NodeId);
+    },
+    [appStateApi],
+  );
+
+  const handleNodeBlur = React.useCallback(() => {
+    appStateApi.blurHoverNode();
+  }, [appStateApi]);
+
+  const handleNodeToggle = React.useCallback(
+    (event: React.SyntheticEvent, nodeIds: string[]) => {
+      setExpandedDomNodeIds(nodeIds);
+    },
+    [setExpandedDomNodeIds],
+  );
+
+  const handleNodeDelete = React.useCallback(
+    (event: React.KeyboardEvent<HTMLUListElement>) => {
+      // delete selected node if event.key is Backspace
+      if (event.key === 'Backspace') {
+        appStateApi.update(
+          (draft) => {
+            if (!selectedDomNodeId) {
```

Comment:
Can you check for if there is a selected node outside of `appStateApi.update`? That way we don't add a new entry to the undo stack, I think.